// convert-pro/convert.js
"use strict";
var __spreadArrays = (this && this.__spreadArrays) || function () {
    for (var s = 0, i = 0, il = arguments.length; i < il; i++) s += arguments[i].length;
    for (var r = Array(s), k = 0, i = 0; i < il; i++)
        for (var a = arguments[i], j = 0, jl = a.length; j < jl; j++, k++)
            r[k] = a[j];
    return r;
};
exports.__esModule = true;
exports.stringToAmountsDegrees = exports.findBest = exports.convertT = exports.convertBet = exports.fix = exports.unitIndexByName = exports.amountToString = exports.dStringToAmounts = exports.convertCreator = exports.force = exports.bytes = exports.degrees = exports.pressure = exports.mass = exports.frequency = exports.energy = exports.temperature = exports.time = exports.volume = exports.area = exports.length = void 0;
/* auto-gen end */
// default options
var dOp = {
    base: 1000,
    lowerCase: false,
    shortcut: true,
    accuracy: 5
};
function higherDim(unitGroup, dim) {
    var result = [];
    var str = dim === 2 ? "Square " : "Cubic ";
    for (var i = 0; i < unitGroup.length; i++) {
        var item = unitGroup[i];
        result.push([str + item[0], str + item[1], item[2] + dim, Math.pow(item[3], dim)]);
    }
    return result;
}
exports.length = [
    ["Femtometer", "Femtometers", "FM", 1e-15, 7],
    ["Picometer", "Picometers", "PM", 1e-12],
    ["Nanometer", "Nanometers", "NM", 1e-9],
    ["Micrometer", "Micrometers", "uM", 1e-6],
    ["Millimeter", "Millimeters", "MM", 0.001],
    ["Centimeter", "Centimeters", "CM", 0.01],
    ["Meter", "Meters", "M", 1],
    ["Kilometer", "Kilometers", "KM", 1000],
    ["Inch", "Inches", "IN", 0.0254],
    ["Foot", "Feet", "FT", 0.3048],
    ["Yard", "Yards", "YD", 0.9144],
    ["Mile", "Miles", "MI", 1609.344],
    ["Nautical Mile", "Nautical Miles", "NMI", 1852],
    ["Astronomical Unit", "Astronomical Units", "AU", 149597870700],
    ["Light Year", "Light Years", "LY", 9460730472580800],
    ["Parsec", "Parsecs", "PC", 30856775814913673],
];
exports.area = __spreadArrays(higherDim(exports.length, 2), [
    ["Acre", "Acres", "Ac", 4046.86],
    ["Deciare", "Deciares", "Da", 10],
    ["Are", "Ares", "Are", 100],
    ["Decare", "Decares", "Decare", 1000],
]);
exports.volume = __spreadArrays(higherDim(exports.length, 3), [
    ["Liter", "Liters", "L", 0.001],
    ["MilliLiter", "MilliLiter", "ML", 1e-6],
    ["Fluid Once", "Fluid Once", "fl-oz", 0.00002841312686461145],
    ["Table Spoon", "Table Spoons", "tbsp", 0.000017758204290382156],
    ["Tea Spoon", "Tea Spoons", "tsp", 0.000005919401430127385],
]);
exports.time = [
    ["Femtosecond", "Femtoseconds", "FS", 1e-15, 11],
    ["Picosecond", "Picoseconds", "PS", 1e-12],
    ["Nanosecond", "Nanoseconds", "NS", 1e-9],
    ["Microsecond", "Microseconds", "uS", 1e-6],
    ["Millisecond", "Milliseconds", "MS", 0.001],
    ["Second", "Seconds", "S", 1],
    ["Minute", "Minutes", "Min", 60],
    ["Hour", "Hours", "H", 3600],
    ["Day", "Days", "H", 86400],
    ["Week", "Weeks", "H", 604800],
    ["Month", "Months", "M", 2628000],
    ["Year", "Years", "Y", 31556952],
    ["Fortnight", "Fortnights", "Fn", 1209600],
    ["Decade", "Decades", "Dec", 315569520],
    ["Century", "Centuries", "Cn", 3155695200],
    ["Millennium", "Millenniums", "Ml", 31556952000],
    ["Shake", "Shakes", "Sk", 1e-8],
    ["Time Unit", "Time Units", "TU", 0.001024],
];
exports.temperature = [
    ["Kelvin", "Kelvin", "K", 1, 0],
    // from cel to kel
    ["Celsius", "Celsius", "C", 1, 273.15],
    ["Fahrenheit", "Fahrenheit", "F", 0.55555556, 255.37222],
    ["Rankine", "Rankine", "R", 0.555556, 0],
    ["Delisle", "Delisles", "De", 0.66666667, 339.816666667],
];
exports.energy = [
    ["Joule", "Joules", "J", 1, 1],
    ["Kilojoule", "Kilojoules", "KJ", 1000],
    ["Calorie", "Calories", "Cal", 4.184, 1],
    ["Kilocalorie", "Kilocalories", "KCal", 4184],
    ["Watt-hour", "Watt-hour", "Wh", 3600, 1],
    ["KiloWatthour", "KiloWatthour", "kWh", 3600000],
    ["Electronvolt", "Electronvolt", "eV", 1.602176634e-19, 4],
    ["Kilo-electronvolt", "Kilo-electronvolt", "KeV", 1.602176634e-16],
    ["Mega-electronvolt", "Mega-electronvolt", "MeV", 1.602176634e-13],
    ["Giga-electronvolt", "Giga-electronvolt", "MeV", 1.602176634e-10],
    ["Tera-electronvolt", "Tera-electronvolt", "TeV", 1.602176634e-7],
    ["British Thermal Unit", "British Thermal Units", "Btu", 1055.06],
    ["Foot-pound", "Foot-pounds", "ft-lb", 1.355818],
];
exports.frequency = [
    ["Hertz", "Hertz", "Hz", 1, 3],
    ["Kilohertz", "kilohertz", "kHz", 1000],
    ["Megahertz", "Megahertz", "mHz", 1000000],
    ["Gigahertz", "Gigahertz", "gHz", 1000000000],
];
exports.mass = [
    ["Femtogram", "Femtograms", "Fg", 1e-15, 6],
    ["Picogram", "Picograms", "Pg", 1e-12],
    ["Nanogram", "Nanograms", "Ng", 1e-9],
    ["Microgram", "Micrograms", "ug", 1e-6],
    ["Milligram", "Milligrams", "mg", 0.001],
    ["Gram", "Grams", "g", 1],
    ["Kilogram", "Kilograms", "Kg", 1000],
    ["Ton", "Tons", "T", 1000000],
    ["Stone", "Stones", "ST", 6350.29],
    ["Pound", "Pounds", "LB", 553.592],
    ["Ounce", "Ounces", "OZ", 28.3495],
];
exports.pressure = [
    ["atmosphere", "atmospheres", "atm", 1],
    ["Bar", "Bars", "bar", 0.986923],
    ["Pascal", "Pascals", "Pa", 0.00000986923],
    ["Pound per square inch", "Pound per square inch", "psi", 0.068046],
    ["torr", "torr", "torr", 0.00131579],
];
exports.degrees = [
    ["degree", "degrees", "°", 1],
    ["arcminute", "arcminutes", "'", 1 / 60],
    ["arcsecond", "seconds", "''", 1 / 3600],
    ["degree", "degrees", "d", 1],
    ["radian", "radians", "r", 180 / Math.PI],
    ["gradian", "gradians", "g", 0.9],
    ["milliradian", "milliradians", "mrad", 180 / Math.PI / 1000],
];
exports.bytes = [
    ["bit", "bits", "bit", 1 / 8],
    ["byte", "bytes", "B", 1],
    ["Kilobyte", "kilobytes", "KB", 1000],
    ["Megabyte", "Megabytes", "MB", Math.pow(1000, 2)],
    ["Gigabyte", "Gigabytes", "GB", Math.pow(1000, 3)],
    ["terabyte", "Terabytes", "TB", Math.pow(1000, 4)],
    ["petabyte", "Petabytes", "PB", Math.pow(1000, 5)],
    ["Kibibyte", "Kibibytes", "KiB", 1024],
    ["Mebibyte", "Mebibytes", "MiB", Math.pow(1024, 2)],
    ["Gibibyte", "Gibibytes", "GiB", Math.pow(1024, 3)],
    ["Tebibyte", "Tebibytes", "TiB", Math.pow(1024, 4)],
    ["Pebibyte", "Pebibytes", "PiB", Math.pow(1024, 5)],
];
exports.force = [
    ["Newton", "Newtons", "N", 1],
    ["Dyne", "Dynes", "Dyn", 1e-5],
    ["Kilogram-force", "Kilogram-force", "KgF", 9.80665],
    ["Pound-force", "Pound-force", "LBF", 4.44822],
    ["Poundal", "Poundals", "Pdl", 0.1382549544],
];
function convertCreator(unitGroup, absoluteValueIndex, convert, findBest, stringToAmounts) {
    if (stringToAmounts === void 0) { stringToAmounts = dStringToAmounts; }
    function converter(from, toOrOptions, maybeOptions) {
        var options = Object.assign({}, dOp, typeof toOrOptions === "object" ? toOrOptions : maybeOptions);
        var to = typeof toOrOptions === "object" ? undefined : toOrOptions;
        if (typeof from === "number") {
            if (!to)
                return amountToString(unitGroup, findBest
                    ? findBest(unitGroup, [from, absoluteValueIndex], options)
                    : [from, absoluteValueIndex], options);
            var unitIndex = unitIndexByName(unitGroup, to);
            var _a = convert(unitGroup, [from, absoluteValueIndex], unitIndex), value = _a[0], index = _a[1];
            if (options.stringify)
                return amountToString(unitGroup, [value, index], options);
            return fix(value, options.accuracy);
        }
        if (typeof from === "string") {
            var arr = stringToAmounts(unitGroup, from);
            var value = 0;
            var toIndex = to ? unitIndexByName(unitGroup, to) : absoluteValueIndex;
            for (var i = 0; i < arr.length; i++) {
                value += convert(unitGroup, arr[i], toIndex)[0];
            }
            if (!(arr.length === 1 && !to) && options.stringify)
                return amountToString(unitGroup, [value, toIndex], options);
            return value;
        }
        if (Array.isArray(from)) {
            if (from.length < 2 || from.length & 1)
                throw new Error("array must be string number pairs and have at least 1 pair");
            var toIndex = to ? unitIndexByName(unitGroup, to) : absoluteValueIndex;
            var result = 0;
            for (var i = 0; i < from.length; i += 2) {
                var value = from[i];
                var unitName = from[i + 1];
                if (!(typeof value === "number" || typeof value === "bigint") ||
                    typeof unitName !== "string")
                    throw new Error("array must be string number pairs and have at least 1 pair");
                result += convert(unitGroup, [Number(value), unitIndexByName(unitGroup, unitName)], toIndex)[0];
            }
            if (options.stringify)
                return amountToString(unitGroup, [result, toIndex], options);
            return result;
        }
        throw new Error("unreachable");
    }
    return converter;
}
exports.convertCreator = convertCreator;
function dStringToAmounts(unitGroup, str) {
    str = str.trim();
    var regex = /([0-9.]+) ([a-zA-Z ]+)/g;
    var arr;
    var result = [];
    var first = false, last = 0;
    while ((arr = regex.exec(str)) !== null) {
        if (!first && arr.index !== 0)
            throw new Error("Invalid String: `" + str + "` at range 0:" + arr.index);
        first = true;
        last = arr.index + arr[0].length;
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        var _ = arr[0], value = arr[1], unitName = arr[2];
        unitName = unitName.trimEnd();
        var unitIndex = unitIndexByName(unitGroup, unitName);
        if (unitIndex === -1)
            throw new Error("Unknown Unit: '" + unitName + "'");
        result.push([Number(value), unitIndex]);
    }
    if (result.length === 0 || last !== str.length)
        throw new Error("Invalid String: `" + str + "` at range " + last + ":" + str.length);
    return result;
}
exports.dStringToAmounts = dStringToAmounts;
function amountToString(unitGroup, amount, options) {
    var _a = unitGroup[amount[1]], name = _a[0], nameWithS = _a[1], shortcut = _a[2];
    var value = options.accuracy ? fix(amount[0], options.accuracy) : amount[0];
    var unitName = lower(options.shortcut ? shortcut : amount[0] > 1 ? nameWithS : name, options.lowerCase);
    return value + " " + unitName;
}
exports.amountToString = amountToString;
function unitIndexByName(unitGroup, name) {
    name = name.toLowerCase();
    var unitIndex = unitGroup.findIndex(function (item) {
        return item[0].toLowerCase() === name ||
            item[1].toLowerCase() === name ||
            item[2].toLowerCase() === name;
    });
    if (unitIndex === -1)
        throw new Error("Unit: '" + name + "' Not Found");
    return unitIndex;
}
exports.unitIndexByName = unitIndexByName;
function lower(str, bool) {
    return bool ? str.toLowerCase() : str;
}
/** function to approximate the number passed to a certain decimal */
function fix(num, decimal) {
    var fixer = Math.pow(10, decimal);
    return Math.round(num * fixer) / fixer;
}
exports.fix = fix;
function convertBet(unitGroup, amount, unitIndex) {
    return [
        amount[0] * (unitGroup[amount[1]][3] / unitGroup[unitIndex][3]),
        unitIndex,
    ];
}
exports.convertBet = convertBet;
function convertT(unitGroup, amount, unitIndex) {
    var unit1 = unitGroup[amount[1]];
    var unit2 = unitGroup[unitIndex];
    var toK = amount[0] * unit1[3] + unit1[4];
    var kToUnit = (toK - unit2[4]) / unit2[3];
    return [kToUnit, unitIndex];
}
exports.convertT = convertT;
function findBestBytes(_, amount, options) {
    var value = amount[0];
    if (value < options.base)
        return [value, 1];
    var finalUnitIndex = options.base === 1000 ? 6 : 11;
    var index = options.base === 1000 ? 2 : 7;
    value = value / options.base;
    while (value >= options.base && index < finalUnitIndex) {
        value = value / options.base;
        index++;
    }
    return [value, index];
}
function findBest(unitGroup, amount) {
    var index;
    for (var i = amount[1]; i >= 0; i--) {
        if (unitGroup[i][4] !== undefined) {
            index = i;
            break;
        }
    }
    if (index === undefined || unitGroup[index][4] + index < amount[1])
        return amount;
    var value = 0, index2 = 0;
    for (var i = index; i <= unitGroup[index][4] + index; i++) {
        value = amount[0] * (unitGroup[amount[1]][3] / unitGroup[i][3]);
        index2 = i;
        var c = unitGroup[i + 1][3] / unitGroup[i][3];
        if (c >= value && Math.abs(c - value) > 0.1)
            return [value, i];
    }
    return [value, index2];
}
exports.findBest = findBest;
function stringToAmountsDegrees(_, str) {
    var dr = /([0-9.]+)[d°](([0-9.]+)')?(([0-9.]+)'')?/g;
    var de = /([0-9.]+)( ?)([a-zA-Z ]+)/g;
    var result = [];
    var arr;
    while ((arr = de.exec(str)) !== null) {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        var _3 = arr[0], value = arr[1], space = arr[2], unitName = arr[3];
        unitName = unitName.trimEnd();
        if (!space && unitName === "d")
            continue;
        var unitIndex = unitIndexByName(exports.degrees, unitName);
        if (unitIndex === -1)
            throw new Error("Unknown Unit: '" + unitName + "'");
        result.push([Number(value), unitIndex]);
    }
    while ((arr = dr.exec(str)) !== null) {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        var _4 = arr[0], value1 = arr[1], _1 = arr[2], value2 = arr[3], _2 = arr[4], value3 = arr[5];
        result.push([Number(value1), 0]);
        if (value2)
            result.push([Number(value2), 1]);
        if (value3)
            result.push([Number(value3), 2]);
    }
    return result;
}
exports.stringToAmountsDegrees = stringToAmountsDegrees;
var convert = {
    temperature: convertCreator(exports.temperature, 0, convertT),
    length: convertCreator(exports.length, 6, convertBet, findBest),
    area: convertCreator(exports.area, 6, convertBet, findBest),
    volume: convertCreator(exports.volume, 6, convertBet, findBest),
    time: convertCreator(exports.time, 5, convertBet, findBest),
    mass: convertCreator(exports.mass, 5, convertBet, findBest),
    energy: convertCreator(exports.energy, 0, convertBet, findBest),
    frequency: convertCreator(exports.frequency, 0, convertBet, findBest),
    pressure: convertCreator(exports.pressure, 0, convertBet),
    degrees: convertCreator(exports.degrees, 0, convertBet, undefined, stringToAmountsDegrees),
    bytes: convertCreator(exports.bytes, 1, convertBet, findBestBytes),
    force: convertCreator(exports.force, 0, convertBet)
};
exports["default"] = convert;
